### ISSUE TITLE
Camera on iOS only available if webview is recreated

### ISSUE THREAD

**Issue by wmaca** [react-native-webview/react-native-webview#2131]
**Bug description:**

With `allowsInlineMediaPlayback` enabled and `mediaPlaybackRequiresUserAction` disabled (this one was a problem on Android), after we ask for camera access on iOS, the camera simply does not load.

After we restart the app, the user is prompted with the webview's request to access the camera (not the SO one, but the one the browser does). And it works.

**To Reproduce:**

1. Render the webview.
2. Send a postMessage to ask native code to request camera permission.
3. Grant permission.
4. Webview's `getMediaDevices()` fail.

**Expected behavior:**
Webview should be able to get media devices.

**Environment:**
 - OS: IOS
 - OS version: 14.6
 - react-native version: 0.63.4
 - react-native-webview version: ^11.2.4

-------

![image](https://user-images.githubusercontent.com/3537591/129076154-ddfea2d5-d68f-4ccf-921f-ee2dc35635b6.png)

**Comment by anhvoduy** [comment]
hi @wmaca : For "mediaPlaybackRequiresUserAction disabled (this one was a problem on Android)" 
=> why don't you try to seperate two web-view: one for Android and one for IOS ? 
Then, configure suitable properties for each web-view

Please refer to document for more detail: https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md

**Comment by wmaca** [comment]
Hi @anhvoduy, we could to that, bu we want `mediaPlaybackRequiresUserAction={false}` on iOS. Any other reason to do that?

**Comment by eslambasher** [comment]
Up, does anyone have a solution please?

**Comment by wmaca** [comment]
If you want a resolution, don't spam asking if there is a solution. Upvote it.

**Comment by wmaca** [comment]
It is still problematic.

**Comment by geiz** [comment]
Referencing
https://github.com/react-native-webview/react-native-webview/issues/1672#issuecomment-938706459

For this workaround, how would we get the permission in the webview before creating the webview?

**Comment by wmaca** [comment]
> Referencing
> 
> https://github.com/react-native-webview/react-native-webview/issues/1672#issuecomment-938706459
> 
> 
> 
> For this workaround, how would we get the permission in the webview before creating the webview?

Not sure I understood your question, as it seems to be self-explanatory. Nevertheless, you would do that by putting any element that is not the webview on the screen until you have the permission (e.g. return isGranted ? <Webview /> : <AskForPermission />).

**Comment by geiz** [comment]
> Not sure I understood your question, as it seems to be self-explanatory. Nevertheless, you would do that by putting any element that is not the webview on the screen until you have the permission (e.g. `return isGranted ? <Webview /> : <AskForPermission />`).

I ran something like this:

```
  const [hasPermissions, setHasPermissions] = useState(false);
...
  useEffect(() => {
    const checkPermissionsWrapper = async () => {
      // No need to check if permissions is already true.
      if (hasPermissions === false) {
        let res = await checkPermissions();
        console.log('hasPermissions triggered: ', res);
        setHasPermissions(res);

        // if no permission we will try asking again
        if (res === false) {
          askForMeetingPermissions();
        }
      }
    };

    checkPermissionsWrapper();
  }, [hasPermissions]);
...
return hasPermissions ? (
    <WebView
      originWhitelist={['*']}
      userAgent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
      source={{uri: 'https://www.onlinemictest.com/webcam-test/'}}
      allowsInlineMediaPlayback={true}
      mediaPlaybackRequiresUserAction={false}
      startInLoadingState={true}
      javaScriptEnabled
      onError={e => console.log('error: ', e)}
      onLoad={data => console.log('load: ', data)}
    />
  ) : (
    <></>
  );```
  
Permissions were obtained and granted with    

Once granted, the webview would render. However, still unable to run the camera or mic. Works fine on android. Could I have some insights as to what is going on?

Packages
    "react-native": "0.66.3",
    "react-native-permissions": "^3.1.0",
    "react-native-webview": "^11.15.0",

**Comment by wmaca** [comment]
I don't, because those versions are not the same I am using.

But, if you app already has permission to use the camera, you should be able to close it and, when opening it again, you should be prompted with browser permission to use the camera (not the native one). If this url has such permission, it should work.

**Comment by hellodoctordev** [comment]
In case anyone's still have this issue, I just resolved my similar problem (video call rendered in the webview wasn't actually playing video or audio) by adding a user agent. Specifically, this one I found on Stack Overflow (though I assume a variety of user agents would have similar results):

`    userAgent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"
`

https://stackoverflow.com/questions/62359050/unable-to-join-live-video-session-react-native-webview-android

**Comment by wmaca** [comment]
This does not seem to be the same problem.

> In case anyone's still have this issue, I just resolved my similar problem (video call rendered in the webview wasn't actually playing video or audio) by adding a user agent. Specifically, this one I found on Stack Overflow (though I assume a variety of user agents would have similar results):
> 
> ` userAgent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36"`
> 
> https://stackoverflow.com/questions/62359050/unable-to-join-live-video-session-react-native-webview-android

**Comment by mawaki** [comment]
@wmaca please check the video in the link below to see if it will help.

https://www.youtube.com/watch?v=GAHWt2HPEIM

**Comment by vinayn-mend** [comment]
@wmaca Got any workarounds or still facing this issue? I have the same problem on Expo managed project

**Comment by doscorch** [comment]
@vinayn-mend I got this working for an expo managed project by using expo's components to ask for permission before rendering the WebView https://docs.expo.dev/versions/v44.0.0/sdk/camera/

`await Camera.requestCameraPermissionsAsync()`

**Comment by AnshulBharatPe** [comment]
Doing the following worked for me.
Add NSCameraUsageDescription, NSMicrophoneUsageDescription in info.plist
Make sure to add these configurations **before loading the url**.

```
let config = WKWebViewConfiguration()
config.userContentController = contentController
config.allowsInlineMediaPlayback = true
if #available(iOS 14.0, *) {
    config.mediaTypesRequiringUserActionForPlayback = []
} else {
    config.requiresUserActionForMediaPlayback = false
}
self.webview = WKWebView(frame: .zero, configuration: config)
self.webview.load(URLRequest(url: url!))
```

If it doesn't work for you. Try to give the permissions before the webview is initialised. 
For me this worked for me even after giving the permissions inside the webview.
